docker/startup/setup-platform.py

```diff
@@ -0,0 +1,249 @@
+                                                       self.config_store[name].get('type', ""))
+
+    @staticmethod
+    def build(identity: str, data: Dict) -> AgentConfig:
+        ac = AgentConfig(identity=identity, **data)
+        return ac
+
+
+@dataclass
+class PlatformConfig:
+    vip_address: str
+    instance_name: str
+
+    verbosity: Optional[str] = "-vv"
+    services: List[AgentConfig] = field(default_factory=list)
+    agents: List[AgentConfig] = field(default_factory=list)
+
+    @staticmethod
+    def build(file: str) -> PlatformConfig:
+        file = Path(file)
+        if not file.exists():
+            raise ValueError("Invalid file specified for PlatformConfig.build")
+
+        contents = yaml.safe_load(file.open().read())
+        pprint(contents)
+
+        def normailze_dashes(data: Dict) -> Dict:
+            new_dict = {}
+            for k, v in data.items():
+                new_k = k.replace('-', '_')
+                new_dict[new_k] = v
+            return new_dict
+
+        contents['config'] = normailze_dashes(data=contents.get('config', {}))
+
+        vip_address = contents['config'].get('vip_address', 'tcp://127.0.0.1:22916')
+        instance_name = contents['config'].get('instance_name', os.environ['HOSTNAME'])
+
+        pc = PlatformConfig(vip_address=vip_address, instance_name=instance_name)
+
+        for identity, service_config in contents.get("services", {}).items():
+            print(service_config)
+            pc.services.append(ServiceConfig.build(identity, service_config))
+
+        for identity, agent_config in contents.get("agents", {}).items():
+            print(identity)
+            pc.agents.append(AgentConfig.build(identity, agent_config))
+
+        pprint(pc.__dict__)
+
+        return pc
+
+
+
+if __name__ == "__main__":
+    
+
+    platform_config_file = os.environ.get("PLATFORM_CONFIG")
+    reinit_platform = os.environ.get("REINITIALIZE")
+    volttron_version = os.environ.get("VOLTTRON_VERSION")
+    pip_cache_dir = os.environ.get("PIP_CACHE_DIR")
+    volttron_home = os.environ.get("VOLTTRON_HOME")
+    
+    env_dir = os.environ.get("VOLTTRON_VENV")
+    if not env_dir:
+        raise ValueError("Invalid $VOLTTRON_VENV directory specified")
+    
+    
+    def get_path_from_home(path: str):
+        """Return a subpath from VOLTTRON_HOME"""
+        return f"{volttron_home}/{path}"
+    
+    
+    def exec(command: List[str]):
+        """Execute a command using Popen and write out stdout and stderr to sys.stdout"""
+        
+        process = subprocess.Popen(cmd, text=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+        while process.poll() is None:
+            line = process.stdout.readline()
+            sys.stdout.write(line)
+            time.sleep(0.1)
+        if process.returncode != 0:
+            sys.exit(process.returncode)
+    
+    # Only create a new virtual env if we need it.
+    if not Path(env_dir).joinpath("bin/python").exists():
```

Comment:
Does this script create a virtual environment for volttron inside a Docker container?

Reply:
Yes, because we want it to be in a specific location so that the site-packages and stuff will work when a user mounts /home/volttron/datavolume directory.

Otherwise they will have to have mounted something in usr or the normal sitepackages and what I tried did not work so well.  This was the cleanest way I could get the requirements done.